implementation/egads/src/main/java/net/opentsdb/query/anomaly/egads/olympicscoring/OlympicScoringNode.java

```diff
@@ -0,0 +1,815 @@
+    case CONFIG:
+      jitter = 0;
+      model_units = null;
+      prediction_start = context.query().startTime().epoch();
+      prediction_intervals = query_time_span / (prediction_interval * 1000);
+      cache_key = null;
+      LOG.info("CONFIG Set cache key to: " + Bytes.pretty(cache_key));
+      break;
+    case EVALUATE:
+    case PREDICT:
+      // TODO - for now, we need the query timespan to be 1h or 1day at the most as
+      // we'll build the model off the start time of the query.
+      long baseline_span = DateTime.parseDuration(config.getBaselinePeriod()) / 1000;
+      if (baseline_span < 86400) {
+        model_units = ChronoUnit.HOURS;
+      } else {
+        model_units = ChronoUnit.DAYS;
+      }
+      
+      jitter = ((OlympicScoringFactory) factory).jitter(context.query(), model_units);
+      TemporalAmount jitter_duration = Duration.ofSeconds(jitter);
+      
+      final TimeStamp start = context.query().startTime().getCopy();
+      final ChronoUnit duration = modelDuration();
+      start.snapToPreviousInterval(1, duration);
+      start.add(jitter_duration);
+      // now snap to ds interval
+      start.snapToPreviousInterval(DateTime.getDurationInterval(ds_interval), 
+          DateTime.unitsToChronoUnit(DateTime.getDurationUnits(ds_interval)));
+      if (start.compare(Op.GT, context.query().startTime())) {
+        start.subtract(model_units == ChronoUnit.HOURS ? 
+            Duration.ofHours(1) : Duration.ofDays(1));
+      }
+      
+      prediction_start = start.epoch();
+      prediction_intervals = (model_units == 
+          ChronoUnit.HOURS ? 3600 : 86400) * 1000 / (prediction_interval * 1000);
+      
+      cache_key = ((OlympicScoringFactory) factory)
+          .generateCacheKey(context.query(), (int) prediction_start);
+      LOG.info("Set cache key to: " + Bytes.pretty(cache_key));
+      break;
+    default:
+      throw new IllegalStateException("Unhandled config mode: " + config.getMode());
+    }
+  }
+  
+  @Override
+  public Deferred<Void> initialize(final Span span) {
+    final class InitCB implements Callback<Void, Void> {
+      @Override
+      public Void call(final Void arg) throws Exception {
+        // trigger the cache lookup.
+        if (cache != null && config.getMode() != ExecutionMode.CONFIG) {
+          cache.fetch(pipelineContext(), cache_key, null)
+            .addCallback(new CacheCB())
+            .addErrback(new CacheErrCB());
+        } else {
+          fetchBaselineData();
+        }
+        return null;
+      }
+    }
+    return super.initialize(span).addCallback(new InitCB());
+  }
+
+  @Override
+  public QueryNodeConfig config() {
+    return config;
+  }
+
+  @Override
+  public void close() {
+    // TODO fire out out what to do here. For now, we'll let the baseline keep
+    // going.
+  }
+
+  @Override
+  public void onNext(final QueryResult next) {
+    if (data_source == null || !data_source.equals(next.dataSource())) {
```

Comment:
I wonder if this will be compiled as one or two `volatile` reads.

Reply:
yeah, it breaks the typically DCLP but I'm not worried too much as the values *should* always be the same. I could correct it with a lock and actually throw an error if the data sources are different.